samples/sample-07/src/test/java/com/example/sample07/NewConsumerRebalanceProtocolTest.java

```diff
@@ -0,0 +1,325 @@
+		CountDownLatch partitionAssignedLatch = new CountDownLatch(1);
+		CountDownLatch listenerLatch = new CountDownLatch(2);
+
+		@KafkaListener(id = GROUP_ID, topics = TOPIC_NAME, autoStartup = "false")
+		void listen(ConsumerRecord<Integer, String> ignored) {
+			listenerLatch.countDown();
+		}
+
+		@Bean
+		KafkaTemplate<Integer, String> template(ProducerFactory<Integer, String> producerFactory) {
+			return new KafkaTemplate<>(producerFactory);
+		}
+
+		@Bean
+		ProducerFactory<Integer, String> producerFactory() {
+			return new DefaultKafkaProducerFactory<>(KafkaTestUtils.producerProps(BROKER));
+		}
+
+		@Bean
+		ConcurrentKafkaListenerContainerFactory<Integer, String>
+		kafkaListenerContainerFactory(ConsumerFactory<Integer, String> consumerFactory) {
+
+			final ConcurrentKafkaListenerContainerFactory<Integer, String> factory =
+					new ConcurrentKafkaListenerContainerFactory<>();
+
+			factory.setConsumerFactory(consumerFactory);
+
+			factory.getContainerProperties().setConsumerRebalanceListener(new ConsumerAwareRebalanceListener() {
+				@Override
+				public void onPartitionsAssigned(Consumer<?, ?> consumer,
+												 Collection<TopicPartition> partitions) {
+					partitionAssignedLatch.countDown();
+				}
+
+				@Override
+				public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+					partitionRevokedLatch.countDown();
+				}
+			});
+			return factory;
+		}
+
+		@Bean
+		ConsumerFactory<Integer, String> consumerFactory() {
+			final Map<String, Object> props = getConsumerProperties();
+			return new DefaultKafkaConsumerFactory<>(props);
+		}
+
+		@Bean
+		public ThreadPoolTaskExecutor threadPoolTaskExecutor() {
+			final ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
+			executor.setCorePoolSize(5);
+			executor.setQueueCapacity(20);
+			executor.setMaxPoolSize(10);
+			return executor;
+		}
+
+		@Bean
+		GenericContainer genericContainer(ConsumerFactory<Integer, String> cf) throws InterruptedException {
+			GenericContainer genericContainer = setUpBroker();
+
+			Properties props = new Properties();
+			props.setProperty(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, BytesDeserializer.class.getName());
+			props.setProperty(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+							  StringDeserializer.class.getName());
+			props.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, BROKER);
+			props.setProperty(ConsumerConfig.GROUP_ID_CONFIG, "hello");
+			props.setProperty(ConsumerConfig.GROUP_PROTOCOL_CONFIG, "consumer");
+			props.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+			Consumer<Object, Object> dummyConsumer = new KafkaConsumer<>(props);
+
+			Thread.sleep(5000);
+			dummyConsumer.close();
+
+			return genericContainer;
+		}
+
+		GenericContainer setUpBroker() {
+			final DockerImageName imageName = DockerImageName.parse(KAFKA_IMAGE_NAME);
+			final GenericContainer broker = new GenericContainer(imageName);
```

Comment:
Why do we use `org.testcontainers:kafka` then if we don't rely on its specific API?